project_task_autocreate/models/project_task_rule.py

```diff
@@ -0,0 +1,150 @@
+        string="On Deletion?",
+    )
+    state_ids = fields.Many2many(
+        string='State',
+        comodel_name='ir.model.state',
+        relation='project_task_autocreate_states_rel',
+    )
+    rule_type = fields.Selection(
+        string='Rule Type',
+        selection=[
+            ('all', 'All conditions must be met (AND)'),
+            ('any', 'At least one condition must be met (OR)')
+        ],
+        default='all',
+        required=True,
+    )
+    template_ids = fields.Many2many(
+        string='Task Templates',
+        comodel_name='project.task.template',
+        relation='project_task_autocreate_rules_rel',
+    )
+    condition_ids = fields.One2many(
+        string='Conditions',
+        comodel_name='project.task.rule.condition',
+        inverse_name='rule_id',
+    )
+    condition_field_id = fields.Many2one(
+        string='Condition Field',
+        comodel_name='ir.model.fields',
+    )
+
+    @api.model
+    def register_model(self, model):
+        self._model_registry |= {model}
+        return True
+
+    @api.model
+    def _selection_model(self):
+        models = self.env['ir.model'].search([
+            ('model', 'in', list(self._model_registry)),
+        ])
+        return [(m.model, m.name) for m in models]
+
+    @api.multi
+    def get_domain(self, records=None):
+        self.ensure_one()
+        domain = self.condition_ids.get_domain(any=(self.rule_type == 'any'))
+        if records:
+            domain.insert(0, ('id', 'in', records.ids))
+        return domain
+
+    @api.onchange('model')
+    def _onchange_model(self):
+        self.condition_ids = False
+        self.state_ids = False
+        self.template_ids = False
+
+    @api.multi
+    def open_condition_wizard(self):
+        self.ensure_one()
+        field_id = self.env.context.get('condition_field_id', False)
+        if field_id:
+            wizard = self.env['project.task.rule.condition.wizard'].create({
+                'field_id': field_id,
+            })
+            return wizard.open_wizard_2()
+        else:
+            wizard = self.env['project.task.rule.condition.wizard']
+            return wizard.open_wizard()
+
+    @api.multi
+    def apply(self, records):
+        for rule in self:
+            to_apply = records.search(rule.get_domain(records))
+            ctx = {'applied_rule_id': rule.id}
+            for rec in to_apply:
+                rule.with_context(ctx).template_ids.create_task(rec)
+        return True
+
+    @api.multi
+    def apply_on_state(self, records, state):
+        state_obj = self.env['ir.model.state'].search([
+            ('model', '=', records._name),
+            ('state', '=', state)
+        ], limit=1)
+        rules = self.filtered(lambda r: state_obj in r.state_ids)
+        return rules.apply(records)
+
+    @api.model
+    def create(self, vals):
+        res = super(ProjectTaskRule, self).create(vals)
+        self.env[res.model]._update_rule_hooks(self.env)
```

Comment:
@jbchouinard I don't think you need to update your classes when adding a rule. This would only work for a non-threaded server. In prefork mode, you are only updating one of your workers. The others will not be affected (theoretically).

Therefore, I would only rely on the delay for each worker to die naturally.